### PR TITLE
Replace bincode with StorageCodec and flexbuffers

### LINKED ISSUE #1411: Remove bincode usage in various places

### PR DESCRIPTION
**Pull request by tillrohrmann**
Replace bincode with StorageCodec and flexbuffers.

This fixes #1411.

### CODE REVIEW

File: crates/types/src/storage.rs [hunk 1]
```diff
@@ -226,3 +191,47 @@ macro_rules! flexbuffers_storage_encode_decode {
...
+    })?);
+    buf.put(&vec[..]);
+    Ok(())
+}
+
+/// Utility method to decode a [`DeserializeOwned`] type from flexbuffers using serde.
+pub fn decode_from_flexbuffers<T: DeserializeOwned, B: Buf>(
+    buf: &mut B,
+) -> Result<T, flexbuffers::DeserializationError> {
+    if buf.remaining() < 4 {
```
Comment (AhmedSoliman) [inline crates/types/src/storage.rs:213]: Any reason against `std::mem::size_of::<u32>()`?

File: crates/types/src/storage.rs [hunk 2]
```diff
@@ -226,3 +191,47 @@ macro_rules! flexbuffers_storage_encode_decode {
...
+pub fn encode_as_flexbuffers<T: Serialize, B: BufMut>(
+    value: T,
+    buf: &mut B,
+) -> Result<(), flexbuffers::SerializationError> {
+    let vec = flexbuffers::to_vec(value)?;
+    // write the length
+    buf.put_u32_le(u32::try_from(vec.len()).map_err(|_| {
+        flexbuffers::SerializationError::custom("only support serializing types of size <= 4GB")
+    })?);
+    buf.put(&vec[..]);
```
Comment (AhmedSoliman) [inline crates/types/src/storage.rs:205]: Can we check that the buffer has enough space? Bytes have the habit of panicking if buffer is small.

**Review by AhmedSoliman — APPROVED** [review]
🥳

File: crates/types/src/storage.rs (hunk 1 above)
Comment (tillrohrmann) [reply crates/types/src/storage.rs:213]: No. Will change it.

File: crates/types/src/storage.rs (hunk 2 above)
Comment (tillrohrmann) [reply crates/types/src/storage.rs:205]: True. Will change it.

**Pull request merged**